Improve text extraction

HarrisonGregg opened the pull request:
Various changes to improve text extraction and a script for extracting text from saved WARC files and adding the extracted text to a CSV.

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+import os
+import sys
+
+import boto3
```
I think you need to add this to `requirements.txt`, right?

HarrisonGregg replied:
Good catch.

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+import magic
+
+from osp_classifier.warc_utils import parse_warc
+from osp_classifier.pipeline_utils import bytes_to_text
```
This should now reference `osp_classifier.text_extraction`.

HarrisonGregg replied:
True

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+        row['Error'] = ""
+
+        try:
+            mime, text = bytes_to_text(doc_bytes)
```
- This line will error because `bytes_to_text` now returns an `error` value as well as `mime` and `text` values.
- I imagine the `error` value should be incorporated into `row['Error']` in some way, right?  Or at least into some type of logging statement?

HarrisonGregg replied:
Fixed

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+from osp_classifier.warc_utils import parse_warc
+from osp_classifier.pipeline_utils import bytes_to_text
+
+csv.field_size_limit(sys.maxsize)
```
Yep.  A few of the `Text` cells go over the normal limit.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+@click.command()
+@click.argument(
+    "csv_in_path",
+    type=click.Path()
```
Sure

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+    type=click.Path()
+)
+@click.argument(
+    "data_root",
```
Given that we're always using the same S3 bucket, I don't think it makes sense to pass in an S3 path, and using an S3 path as a file path was only possible in `osp-scraper` because of some `FilesPipeline` magic.  So instead I just made it so you can pass "s3" as the file path and have it download the files from the bucket specified in the settings.

jodizzle replied:
Sure, that sounds fine for now.  We can generalize the bucket name later if we want to.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+    "--retry",
+    default=False,
+    is_flag=True,
+    help="Try to extract text for rows without extracted text"
```
Done

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+    new_fields = [
+        'MIME',
+        'Text',
+        'ExtractedLength',
```
Sounds good

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+
+    i = 0
+    for row in csv_reader:
+        print(i)
```
Agreed

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+    i = 0
+    for row in csv_reader:
+        print(i)
+        i += 1
```
Yep

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+        path = data_root + "/" + key
+        if download_missing and not os.path.isfile(path):
+            os.makedirs(os.path.dirname(path), exist_ok=True)
+            s3.meta.client.download_file("syllascrape", key, path)
```
I've made it into a setting loaded from an environment variable (but with default "syllascrape")

jodizzle replied:
That sounds like a good idea.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+        for field in new_fields:
+            row.setdefault(field, "")
+
+        row['MIME'] = magic.from_buffer(doc_bytes or b'', mime=True)
```
- No, it shouldn't be
 - If I use the MIME from `bytes_to_text`, unfortunately the MIME doesn't get set if the text extraction raises an exception, so I do it separately

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,101 @@
+            row['Text'] = text or ""
+        except Exception as e:
+            row['Error'] = str(e)
+            print("Error", str(e))
```
Changed to a logging statement

HarrisonGregg commented on osp_classifier/text_extraction.py:
```diff
@@ -30,16 +29,26 @@ def call_with_bytes(b, command, *args):
+            f"Extraction failed with subprocess error: {error.decode()}"
+        )
+
+    return result.decode('utf-8'), error.decode()
```
At one point I had `result.decode('utf-8', 'ignore')`, and I guess I forgot the 'utf-8' when I removed the 'ignore'.

HarrisonGregg commented on osp_classifier/text_extraction.py:
```diff
@@ -94,20 +96,24 @@ def bytes_to_text(b):
+        text, error = rtf_to_text(b)
 
-    return mime, text
+    # Remove null bytes, which are unnecessary and can cause problems
```
Fixed

HarrisonGregg commented:
I think I addressed all your comments, and in the process refactored the `extract_text.py` script to use a pandas dataframe to be more consistent with the rest of the repo.

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+            warc_data = s3.meta.client.download_fileobj(s3_bucket, key)
+            warc_file = parse_warc(warc_data)
+        else:
+            path = data_root + "/" + key
```
Using `os.path.join` is probably the safer choice.

HarrisonGregg replied:
Fixed

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+                os.makedirs(os.path.dirname(path), exist_ok=True)
+                s3.meta.client.download_file(s3_bucket, key, path)
+            with open(path, "rb") as warc_data:
+                warc_file = parse_warc(warc_data)
```
So in the case that `path` isn't available, do you want to handle the missing file error in some way?  As in, should `data_root` have to contain all the keys described in `csv_in` (and the user be expected to pass `--download_missing` if not), or should there be some more leniency?

HarrisonGregg replied:
I've updated it to log an info message when it's downloading a file and a warning when there's a missing file, but to skip the file without failing.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,100 @@
+def main(csv_in, csv_out, data_root, retry, download_missing, chunksize):
+    """Extract text from WARC files as additional columns in new CSV.
+
+    \b
```
Yeah.  It prevents rewrapping of the following paragraph.  Normally I'd add a comment for something like that in code, but you can't really do that here

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+
+        return {
+            'Text': (text or ""),
+            'Error': (error or "")
```
I guess not.  I had them on one line at first and it was easier to read that way, but I'll remove them now.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+            df = df[df['Text'].eq("")]
+
+        doc_bytes = df['Key'].map(load_warc_body)
+        # Ensure that MIME is set even if text extraction fails
```
Fixed

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+        logger.info(f"{total_written} entries written to CSV")
+
+
+
```
Fixed

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+
+    def load_warc_body(key):
+        if data_root.lower() == "s3":
+            warc_data = s3.meta.client.download_fileobj(s3_bucket, key)
```
Oh right.  Fixed.

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,100 @@
+      Path to directory with saved WARC files, or 'S3'.
+
+    To try to extract text again for rows that have no extracted text, use a
+    previous csv_out as the csv_in and use the '--retry' flag.
```
Fixed

HarrisonGregg commented on bin/extract_text.py:
```diff
@@ -0,0 +1,102 @@
+@click.command()
+@click.argument("csv_in", type=click.Path(exists=True))
+@click.argument("csv_out", type=click.Path())
+@click.argument("data_root", type=click.Path(exists=True))
```
Fixed

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,108 @@
+        total_loaded += len(df)
+        logger.info(f"{total_loaded} entries loaded from CSV")
+        if retry:
+            df = df[df['Text'].eq("")]
```
Using `--retry` seems to currently `KeyError`.  I think you mean to put `df = df[~df['Text'].eq("")]`?

jodizzle replied:
Oops, my mistake.  The `KeyError` came from an inappropriate input CSV and had nothing to do with this line.

jodizzle commented on bin/extract_text.py:
```diff
@@ -0,0 +1,108 @@
+        df['Text'] = results['Text']
+        df['Error'] = results['Error']
+        df['TextLength'] = df['Text'].map(len)
+        df.to_csv(csv_out, mode="a" if i else "w", header=(not i), index=False)
```
For reference, I think our discussion in https://github.com/opensyllabus/osp-classifier/pull/4#discussion_r146095595 is relevant here as well.

HarrisonGregg replied:
I guess doing it with a file instead of a path makes more sense here.

jodizzle commented on osp_classifier/text_extraction.py:
```diff
@@ -1,6 +1,5 @@
-import docx
 import re
 import subprocess
 import magic
```
This should be grouped with the rest of the third-party imports.

jodizzle commented on osp_classifier/text_extraction.py:
```diff
@@ -30,16 +29,26 @@ def call_with_bytes(b, command, *args):
-    result = proc.communicate(input=b, timeout=10)
+    try:
+        result, error = proc.communicate(input=b, timeout=10)
+    except TimeoutExpired as e:
```
I think you need `subprocess.TimeoutExpired` here.

HarrisonGregg replied:
Done

jodizzle commented on osp_classifier/text_extraction.py:
```diff
@@ -30,16 +29,26 @@ def call_with_bytes(b, command, *args):
+    if proc.returncode:
+        raise Exception(
+            f"Extraction failed with subprocess error: {error.decode()}"
+        )
```
- Should we raise a more descriptive Exception (and/or make our own)?
- If the `TimeoutExpired` error is caught in the earlier lines, then there will be a `proc.returncode`, but there's a chance that `error.decode()` is the empty string.  Given that, it might be a good idea to include in the exception string some extra info, like the `proc.returncode`.

HarrisonGregg replied:
Yeah, I think including the return code makes sense.

jodizzle commented:
Is `text_extraction.unbreak_text` used for anything?  Should it be?
